### PR TITLE
Work on new payment channels construction

### PR DESCRIPTION
**Pull request by whyrusleeping**
This gets us closer to a payment channel setup that we can actually use for paying storage miners. It also has a few other features that will be important for efficiency.

### CODE REVIEW

**Comment by pooja** [comment]
Hey @whyrusleeping lots of great conversation and updates here. Did you want review on this from someone else too (before merging)?

**Comment by whyrusleeping** [comment]
@pooja review from nicola would be good. I also need to add in another payment conditional. We need to be able to pay contingent on a particular file inclusion proof linking user data to a sector.

**Comment by whyrusleeping** [comment]
Alright, this is a good step forward. Gonna merge for now, and follow up on additional questions soon.

File: payments.md [hunk 1]
```diff
@@ -26,67 +26,7 @@ For example:
...
-```
-
-Channel IDs should be memory efficient and generated is such a way that reordering does not change the channelID. This may be solved by first indexing by the target address internally and then using the nonce of the message that invoked the create channel method.
-
-
-
-### Multi-Lane Payment Channel (WIP)
+### Multi-Lane Payment Channel
 
 The filecoin storage market may require a way to do incremental payments between two parties, over time, for multiple different transactions. The primary motivating usecase for this is to provide payment for file storage over time, for each file stored. An additional requirement is the ability to have less than one message on chain per transaction 'lane', meaning that payments for multiple files should be aggregateable (Note: its okay if this aggregation is an interactive process).
```
Comment (nicola) [inline payments.md]: what does interactive process means here?

File: payments.md [hunk 2]
```diff
@@ -110,16 +47,173 @@ type SpendVoucher struct {
...
+    Redeemed TokenAmount
+}
+
+func (paych *PaymentChannel) validateSignature(sv SpendVoucher) {
+    if msg.From == paych.From {
+        ValidateSignature(sv.SerializeNoSig(), sv.Signature, paych.To)
+    } else if msg.From == paych.To {
+        ValidateSignature(sv.SerializeNoSig(), sv.Signature, paych.From)
+    } else {
+        Fatal("bad programmer")
```
Comment (nicola) [inline payments.md]: uhmm..

File: payments.md [hunk 3]
```diff
@@ -110,16 +47,173 @@ type SpendVoucher struct {
...
+    }
+}
+
+func (paych *PaymentChannel) UpdateChannelState(sv SpendVoucher, secret []byte, pip *PieceInclusionProof) {
```
Comment (nicola) [inline payments.md]: I am not sure I like the idea of hard-coding the `pip` here. In fact, payment channels should be completely agnostic from the rest of filecoin.

In other words, we could rename this payment channel actor the retrieval market payment channel, but that would be a bad idea.

Proposal: define the method of another contract that should be called and return true in order to unclock the funds, in other words, making fund unlocking before timeout to be contingent on another contract execution.

File: payments.md [hunk 4]
```diff
@@ -110,16 +47,173 @@ type SpendVoucher struct {
...
+        Fatal("cannot use this voucher yet!")
+    }
+
+    if sv.SecretPreimage != nil {
+        if Hash(secret) != sv.SecretPreimage {
+            Fatal("Incorrect secret!")
+        }
+    }
+  
+  	if sv.DataCommitment != nil {
```
Comment (nicola) [inline payments.md]: see my previous comment, my opinion is that either this contract is renamed or it should be more generic.

File: payments.md [hunk 5]
```diff
@@ -110,16 +47,173 @@ type SpendVoucher struct {
...
+      // Checks that the piece inclusion proof is valid, and that the referenced sector
+      // is correctly being stored
+      if !ValidateInclusion(pip, sv.DataCommitment) {
```
Comment (nicola) [inline payments.md]: Is this defined anywhere?

File: payments.md [hunk 6]
```diff
@@ -110,16 +47,173 @@ type SpendVoucher struct {
...
+        }
+        if paych.MinCloseHeight < sv.MinCloseHeight {
+            paych.MinCloseHeight = sv.MinCloseHeight
+        }
+    }
+}
+
+func (paych *PaymentChannel) Withdraw(upTo TokenAmount) {
+    // TODO: this ones tricky, withdraw funds without closing it out entirely...
+}
```
Comment (nicola) [inline payments.md]: is this the same as collect?

File: payments.md [hunk 7]
```diff
@@ -110,16 +47,173 @@ type SpendVoucher struct {
...
+        ValidateSignature(sv.SerializeNoSig(), sv.Signature, paych.To)
+    } else if msg.From == paych.To {
+        ValidateSignature(sv.SerializeNoSig(), sv.Signature, paych.From)
+    } else {
+        Fatal("bad programmer")
+    }
+}
+
+func (paych *PaymentChannel) UpdateChannelState(sv SpendVoucher, secret []byte, pip *PieceInclusionProof) {
+    if !paych.validateSignature(sv) {
```
Comment (dignifiedquire) [inline payments.md]: where is `validateSignature` defined?

File: payments.md [hunk 8]
```diff
@@ -110,16 +47,173 @@ type SpendVoucher struct {
...
+        Fatal("Signature Invalid")
+    }
+
+    if chain.Now() < sv.TimeLock {
+        Fatal("cannot use this voucher yet!")
+    }
+
+    if sv.SecretPreimage != nil {
+        if Hash(secret) != sv.SecretPreimage {
```
Comment (dignifiedquire) [inline payments.md]: what `Hash`?

File: payments.md [hunk 9]
```diff
@@ -110,16 +47,173 @@ type SpendVoucher struct {
...
+  	if sv.DataCommitment != nil {
+      // Checks that the piece inclusion proof is valid, and that the referenced sector
+      // is correctly being stored
+      if !ValidateInclusion(pip, sv.DataCommitment) {
+        Fatal("PieceInclusionProof was invalid")
+      }
+  	}
+
+    if sv.RequiredSector != nil {
+        miner, found := GetMiner(msg.From)
```
Comment (dignifiedquire) [inline payments.md]: Where is `GetMiner` defined ?

File: payments.md [hunk 10]
```diff
@@ -110,16 +47,173 @@ type SpendVoucher struct {
...
+        if !found {
+            Fatal("Redeemer is not a miner")
+        }
+
+        if !miner.HasSector(sv.RequiredSector) {
```
Comment (dignifiedquire) [inline payments.md]: Where is `HasSector` defined ?

File: payments.md (hunk 2 above)
Comment (whyrusleeping) [reply payments.md]: can change to  "illegal state"

File: payments.md (hunk 7 above)
Comment (dignifiedquire) [reply payments.md]: I am blind..

File: payments.md (hunk 6 above)
Comment (whyrusleeping) [reply payments.md]: probably going to delete this one anyways. its difficult to do right, and we can just close channels out

**Pull request merged**